src/Build/Graph/ProjectInterpretation.cs

```diff
@@ -0,0 +1,387 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Globalization;
+using System.Linq;
+using Microsoft.Build.BackEnd;
+using Microsoft.Build.Collections;
+using Microsoft.Build.Evaluation;
+using Microsoft.Build.Exceptions;
+using Microsoft.Build.Execution;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.Experimental.Graph
+{
+    internal sealed class ProjectInterpretation
+    {
+        private const string FullPathMetadataName = "FullPath";
+        private const string ToolsVersionMetadataName = "ToolsVersion";
+        private const string SetConfigurationMetadataName = "SetConfiguration";
+        private const string SetPlatformMetadataName = "SetPlatform";
+        private const string SetTargetFrameworkMetadataName = "SetTargetFramework";
+        private const string GlobalPropertiesToRemoveMetadataName = "GlobalPropertiesToRemove";
+        private const string ProjectReferenceTargetIsOuterBuildMetadataName = "OuterBuild";
+
+        private static readonly char[] PropertySeparator = MSBuildConstants.SemicolonChar;
+
+        public static ProjectInterpretation Instance = new ProjectInterpretation();
+
+        private ProjectInterpretation()
+        {
+        }
+
+        private static readonly ImmutableList<GlobalPropertiesModifier> ModifierForNonMultitargetingNodes = new[] {(GlobalPropertiesModifier) ProjectReferenceGlobalPropertiesModifier}.ToImmutableList();
+
+        internal enum ProjectType
+        {
+            OuterBuild, InnerBuild, NonMultitargeting
+        }
+
+        public IEnumerable<ConfigurationMetadata> GetReferences(ProjectInstance requesterInstance)
+        {
+            IEnumerable<ProjectItemInstance> references;
+            IEnumerable<GlobalPropertiesModifier> globalPropertiesModifiers = null;
+
+            switch (GetProjectType(requesterInstance))
+            {
+                case ProjectType.OuterBuild:
+                    references = GetInnerBuildReferences(requesterInstance);
```

Comment:
We should find a place to write down the implication of this: an outer build **cannot** have its own distinct `ProjectReference`s. That's the case today, but I could see a situation where you might want it: a packaging project that packs its inner build outputs + "other non-TF-specific stuff" acquired through reference).

I don't think this is a bad limitation. Just worth formalizing.